codeguru_profiler_agent/profiler_runner.py

```diff
@@ -71,27 +71,40 @@ def _profiling_command(self):
             if self._first_execution:
                 self.collector.setup()
                 self._first_execution = False
-            return self._run_profiler()
+            sample_result = self._run_profiler()
+            if sample_result.success and sample_result.should_check_overall:
+                if self.profiler_disabler.should_stop_profiling(profile=self.collector.profile):
+                    return False
+                if sample_result.should_reset:
+                    self.collector.reset()
+                    return True
+            return sample_result.success
         except:
             logger.info("An unexpected issue caused the profiling command to terminate.", exc_info=True)
             return False
 
     @with_timer("runProfiler")
     def _run_profiler(self):
-        if self.profiler_disabler.should_stop_profiling(self.collector.profile):
-            return False
+        if self.profiler_disabler.should_stop_sampling(self.collector.profile):
+            return RunProfilerStatus(success=False, should_check_overall=False, should_reset=False)
 
+        refreshed_config = False
         if not self.is_profiling_in_progress:
             self._refresh_configuration()
+            refreshed_config = True
 
         # after the refresh we may be working on a profile
         if self.is_profiling_in_progress:
-            if self.collector.flush():
+            if self.collector.flush(reset=False):
                 self.is_profiling_in_progress = False
-                return True
-            sample = self.sampler.sample()
-            self.collector.add(sample)
-        return True
+                return RunProfilerStatus(success=True, should_check_overall=True, should_reset=True)
+            self._sample_and_aggregate()
+        return RunProfilerStatus(success=True, should_check_overall=refreshed_config, should_reset=False)
```

Comment:
Unless I am mistaken, this will make us check for the overall overhead twice per cycle, after we refresh config and after we report instead of just after we report. Why not return  `should_check_overall=False` here and drop the `refreshed_config` variable?

Reply:
Because I think we should stop the agent if the overall cpu_usage for calling configure_agent is high as well. The check is cheap anyway so I went for "better to be safe than sorry" approach :)

Reply:
But unfortunately this check for "overall" overhead makes sense only when we have data points for all the times we called `runProfiler`, most of the time it only samples which should have little overhead and sometimes it calls agent config or reports which we know takes more cpu.

Just before we call the agent configuration we should have cleared the previous data points which means if you run the check right after you will only consider the overhead of calling agent config which we know can be high. Actually with current implementation of the check, it will always return False because of the `MINIMUM_MEASURES_IN_DURATION_METRICS` and at that time we have only 1 data point.